experimental-module-builds -Slint compiler produces invalid Rust when using 'enums'

### Bug Description

Using an enum in a specific combination will make the Slint compiler to produce invalid Rust code:
```
error[E0308]: mismatched types
    --> /home/benny/dev/slint-enum-bug/target/debug/build/gallery-fb6b1ef78d7639e9/out/app-window.rs:1963:255
     |
1963 | ... (((((_self . parent . upgrade () . as_ref () . map (| x | (InnerCosiumComponent_root_28 :: FIELD_OFFSETS . r#root_28_demo_prop_out ()) . apply_pin (x . as_pin_ref ())) . map (| x | x . get ()) . unwrap_or_default ()) . clone ())) == (((((InnerComponent_rectangle_29 :: FIELD_OFFSETS . r#model_data ()) . apply_pin (_self) . get ()) . r#coso1) . clone ())))) ...
     |       -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `slint_generatedAppWindow::DemoType`, found `demo_library::demo_library::DemoType`
     |       |
     |       expected because this is `slint_generatedAppWindow::DemoType`
     |
     = note: `demo_library::demo_library::DemoType` and `slint_generatedAppWindow::DemoType` have similar names, but are actually distinct types
note: `demo_library::demo_library::DemoType` is defined in crate `demo_library`
    --> /home/benny/dev/slint-enum-bug/target/debug/build/demo-library-94d1105d83ba4601/out/demo-library.rs:8:84
     |
   8 |      # [allow (dead_code)] # [derive (Default , Copy , Clone , PartialEq , Debug)] pub enum r#DemoType {
     |                                                                                    ^^^^^^^^^^^^^^^^^^^
note: `slint_generatedAppWindow::DemoType` is defined in the current crate
    --> /home/benny/dev/slint-enum-bug/target/debug/build/gallery-fb6b1ef78d7639e9/out/app-window.rs:9:84
     |
   9 |      # [allow (dead_code)] # [derive (Default , Copy , Clone , PartialEq , Debug)] pub enum r#DemoType {
     |                                                                                    ^^^^^^^^^^^^^^^^^^^

error[E0605]: non-primitive cast: `demo_library::demo_library::DemoType` as `slint_generatedAppWindow::DemoType`
    --> /home/benny/dev/slint-enum-bug/target/debug/build/gallery-fb6b1ef78d7639e9/out/app-window.rs:1972:227
     |
1972 | ... (| x | x . set (((InnerComponent_rectangle_29 :: FIELD_OFFSETS . r#model_data ()) . apply_pin (_self) . get ()) . r#coso1 as _)) ;
     |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object

```

The 'DemoType' enum belong the 'demo-library' crate, and when used should just be imported to the 'demo-app' by a 'use ...' statement. Instead a duplicate version is created by the Slint compiler, which causes this Rust compiler error.

Here is a simple project that reproduces this error:
[slint-enum-bug.zip](https://github.com/user-attachments/files/28172927/slint-enum-bug.zip)

### Reproducible Code (if applicable)

```slint

```

### Environment Details

- Slint Version:
- Platform/OS:
- Programming Language:
- Backend/Renderer:


### Product Impact

_No response_

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -74,6 +74,8 @@ members = [
   'tests/manual/module-builds/blogica',
   'tests/manual/module-builds/blogicb',
   'tests/manual/module-builds/app',
+  'tests/issues/test-11836-project/test-app',
+  'tests/issues/test-11836-project/test-library',
   'tests/backends/',
   'tools/compiler',
   'docs/slint-doc-generator',

--- a/internal/compiler/passes/collect_structs_and_enums.rs
+++ b/internal/compiler/passes/collect_structs_and_enums.rs
@@ -26,16 +26,18 @@ pub fn collect_structs_and_enums(doc: &Document) {
     used_types.structs_and_enums = Vec::with_capacity(hash.len());
     while let Some(next) = hash.iter().next() {
         let key = next.0.clone();
-        if let Some(library_info) = doc.library_exports.get(key.as_str()) {
-            // This is a type imported from an external library, just skip it for code generation
-            hash.remove(&key);
-            used_types.library_types_imports.push((key, library_info.clone()));
-            continue;
-        }
 
-        // Here, using BTreeMap::pop_first would be great when it is stable
-        let used_struct_and_enums = &mut used_types.structs_and_enums;
-        sort_types(&mut hash, used_struct_and_enums, &key);
+        sort_types(&mut hash, &key, &mut |key, ty| {
+            // Here, using BTreeMap::pop_first would be great when it is stable
+            let used_struct_and_enums = &mut used_types.structs_and_enums;
+
+            if let Some(library_info) = doc.library_exports.get(key.as_str()) {
+                // This is a type imported from an external library, just skip it for code generation
+                used_types.library_types_imports.push((key.clone(), library_info.clone()));
+            } else {
+                used_struct_and_enums.push(ty.clone());
+            }
+        });
     }
 }
 
@@ -66,16 +68,20 @@ fn collect_types_in_component(root_component: &Rc<Component>, hash: &mut BTreeMa
 
 /// Move the object named `key` from hash to vector, making sure that all object used by
 /// it are placed before in the vector
-fn sort_types(hash: &mut BTreeMap<SmolStr, Type>, vec: &mut Vec<Type>, key: &str) {
+fn sort_types(
+    hash: &mut BTreeMap<SmolStr, Type>,
+    key: &SmolStr,
+    visitor: &mut impl FnMut(&SmolStr, &Type),
+) {
     let ty = if let Some(ty) = hash.remove(key) { ty } else { return };
     if let Type::Struct(s) = &ty
         && let StructName::User { .. } = &s.name
     {
         for sub_ty in s.fields.values() {
-            visit_declared_type(sub_ty, &mut |name, _| sort_types(hash, vec, name));
+            visit_declared_type(sub_ty, &mut |name, _| sort_types(hash, name, visitor));
         }
     }
-    vec.push(ty)
+    visitor(key, &ty)
 }
 
 /// Will call the `visitor` for every named struct or enum that is not builtin

--- a/tests/issues/test-11836-project/test-app/Cargo.toml
+++ b/tests/issues/test-11836-project/test-app/Cargo.toml
@@ -1,0 +1,17 @@
+# Copyright © Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+# SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+[package]
+name = "issue-11836-test-app"
+version.workspace = true
+edition.workspace = true
+authors = ["Slint Developers <info@slint.dev>"]
+publish = false
+license = "GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0"
+
+[dependencies]
+slint = { path = "../../../../api/rs/slint" }
+issue-11836-test-library = { path = "../test-library" }
+
+[build-dependencies]
+slint-build = { path = "../../../../api/rs/build", features = ["experimental-module-builds"] }

--- a/tests/issues/test-11836-project/test-app/build.rs
+++ b/tests/issues/test-11836-project/test-app/build.rs
@@ -1,0 +1,7 @@
+// Copyright © Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+fn main() {
+    let config = slint_build::CompilerConfiguration::new();
+    slint_build::compile_with_config("ui/app-window.slint", config).expect("Slint build failed");
+}

--- a/tests/issues/test-11836-project/test-app/src/main.rs
+++ b/tests/issues/test-11836-project/test-app/src/main.rs
@@ -1,0 +1,14 @@
+// Copyright © Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+use issue_11836_test_library::test_library;
+use slint::ComponentHandle;
+
+slint::include_modules!();
+
+fn main() {
+    let ui = AppWindow::new().unwrap();
+
+    test_library::init(&ui);
+    ui.run().unwrap();
+}

--- a/tests/issues/test-11836-project/test-app/ui/app-window.slint
+++ b/tests/issues/test-11836-project/test-app/ui/app-window.slint
@@ -1,0 +1,12 @@
+// Copyright © Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+import {
+    TestComponent,
+    TestGlobal,
+} from "@test_library";
+export { TestGlobal }
+
+export component AppWindow inherits Window {
+    TestComponent { }
+}

--- a/tests/issues/test-11836-project/test-library/Cargo.toml
+++ b/tests/issues/test-11836-project/test-library/Cargo.toml
@@ -1,0 +1,17 @@
+# Copyright © Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+# SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+[package]
+name = "issue-11836-test-library"
+version.workspace = true
+edition.workspace = true
+authors = ["Slint Developers <info@slint.dev>"]
+publish = false
+license = "GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0"
+links = "test_library"
+
+[dependencies]
+slint = { path = "../../../../api/rs/slint" }
+
+[build-dependencies]
+slint-build = { path = "../../../../api/rs/build", features = ["experimental-module-builds"] }

--- a/tests/issues/test-11836-project/test-library/build.rs
+++ b/tests/issues/test-11836-project/test-library/build.rs
@@ -1,0 +1,12 @@
+// Copyright © Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+fn main() -> Result<(), slint_build::CompileError> {
+    let config = slint_build::CompilerConfiguration::new()
+        .as_library("test_library")
+        .rust_module("test_library");
+
+    slint_build::compile_with_config("ui/test-library.slint", config)?;
+
+    Ok(())
+}

--- a/tests/issues/test-11836-project/test-library/src/lib.rs
+++ b/tests/issues/test-11836-project/test-library/src/lib.rs
@@ -1,0 +1,14 @@
+// Copyright © Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+pub mod test_library {
+    slint::include_modules!();
+    use slint::ComponentHandle;
+
+    pub fn init<T>(_ui: &T)
+    where
+        T: ComponentHandle + 'static,
+        for<'a> TestGlobal<'a>: slint::Global<'a, T>,
+    {
+    }
+}

--- a/tests/issues/test-11836-project/test-library/ui/test-library.slint
+++ b/tests/issues/test-11836-project/test-library/ui/test-library.slint
@@ -1,0 +1,75 @@
+// Copyright © Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-2.0 OR LicenseRef-Slint-Software-3.0
+
+import { Button } from "std-widgets.slint";
+
+export enum TestType {
+    test1,
+    test2,
+    test3,
+    test4,
+}
+
+export enum TestAction {
+    test1,
+    test2,
+    test3,
+    test4,
+}
+
+export enum TestType2 {
+    test1,
+    test2,
+    test3,
+    test4,
+}
+
+export struct TestActions {
+    test1: TestAction,
+    test2: TestAction,
+}
+
+export struct TestStruct {
+    test1: TestType,
+    test2: bool,
+}
+
+export global TestGlobal {
+    in-out property <[TestStruct]> test-in-out;
+    callback test-callback(i1: int, s1: TestStruct);
+}
+
+export component TestSubComponent inherits Rectangle {
+    out property <TestType> demo-prop-out: TestType.test1;
+    in-out property <[TestStruct]> test-in-out <=> TestGlobal.test-in-out;
+
+    for btn-config[index] in test-in-out: Rectangle {
+        Button {
+            property <TestType> test-prop: btn-config.test1;
+
+            checked: (demo-prop-out == self.test-prop);
+            clicked => {
+                root.demo-prop-out = self.test-prop;
+            }
+        }
+    }
+}
+
+struct TestStruct1 {
+    test1: TestType,
+    test2-actions: TestActions,
+}
+
+struct TestStruct2 {
+    test-map: [[TestStruct1]],
+}
+
+export component TestComponent {
+    in-out property <TestStruct2> test-configs;
+
+    function do-test-test(test1: TestType) {
+        test-configs.test-map[0][0].test1 = test1;
+    }
+
+    TestSubComponent { }
+}

